ldpd/lde.c

```diff
@@ -1759,24 +1768,40 @@ static void zclient_sync_init(unsigned short instance)
 	zclient_sync->session_id = 1; /* Distinguish from main session */
 	zclient_sync->privs = &lde_privs;
 
-	while (zclient_socket_connect(zclient_sync) < 0) {
+	if (zclient_socket_connect(zclient_sync) < 0) {
 		log_warnx("Error connecting synchronous zclient!");
-		sleep(1);
+		goto retry;
 	}
 	/* make socket non-blocking */
 	sock_set_nonblock(zclient_sync->sock);
 
 	/* Send hello to notify zebra this is a synchronous client */
-	while (zclient_send_hello(zclient_sync) < 0) {
+	if (zclient_send_hello(zclient_sync) < 0) {
 		log_warnx("Error sending hello for synchronous zclient!");
-		sleep(1);
+		goto retry;
 	}
 
 	/* Connect to label manager */
-	while (lm_label_manager_connect(zclient_sync, 0) != 0) {
+	if (lm_label_manager_connect(zclient_sync, 0) != 0) {
 		log_warnx("Error connecting to label manager!");
-		sleep(1);
+		goto retry;
 	}
+
+	/* Finish init once the LM session is running */
+	lde_label_list_init();
+
+	return;
+
+retry:
+
+	/* Discard failed zclient object */
+	zclient_stop(zclient_sync);
+	zclient_free(zclient_sync);
+	zclient_sync = NULL;
+
+	/* Retry using a timer */
+	thread_add_timer(master, zclient_sync_retry,
+			 (void *)(intptr_t)instance, 1, NULL);
```

Comment:
Now that we have the `session_id` field, ldpd doesn't need the instance hack anymore. I'd appreciate if you could remove it as part of this PR in order to simplify the code.

Reply:
Do you mean "just send a zero" for the instance value?

Reply:
Yes. ldpd doesn't support multiple instances and probably never will.

Reply:
That was added specifically for the LM instance hack :)

Even bgpd has `-I, --int_num`, which clearly isn't necessary anymore as well.

Reply:
Sorry guys, but this will not solve the problem. There is another loop line 1827 in lde_label_list_init() function which is call just right after zclient_sync_init() function line 177. If zclient_sync_init() failed, then call to lde_get_label_chunk () line 1827  will failed and will enter in the same kind of loop.

In fact, like for OSPF-SR or IS-IS-SR, if we can't get connection with the Label Manager, we can't start LDP. For OSPF-SR and IS-IS-SR, I remove the loop and return a failure and doesn't start Segment Routing. I'm wondering if we should do the same here. Without connection to the Label Manager, LDP has no valid label, thus can't start properly. The other solution is to go to Asynchronous connection with Label Manager and trigger the real start of LDP once getting the first label chunk.